vmf/scripts/mods/vmf/modules/core/packages.lua

```diff
@@ -0,0 +1,129 @@
+  end
+
+  if self:has_package_loaded(package_name) then
+    self:error("Package '%s' has already been loaded", package_name)
+    return
+  end
+
+  local mod_handle = self:get_internal_data("mod_handle")
+
+  if not mod_handle then
+    self:error("Failed to get mod handle. Package management is not available.")
+    return
+  end
+
+  if not _loaded_packages[self] then
+    _loaded_packages[self] = {}
+  end
+
+  local resource_package = Mod.resource_package(mod_handle, package_name)
+
+  if not resource_package then
+    self:error("Could not find package '%s'.", package_name)
+    return
+  end
+
+  local is_loading = self:is_package_loading(package_name)
+
+  if not callback then
+    if not is_loading then
+      resource_package:load()
+    end
+
+    resource_package:flush()
+
+    _loaded_packages[self][package_name] = resource_package
+  else
+    if is_loading then
+      self:error("Package '%s' is currently loading", package_name)
+      return
+    end
+
+    table.insert(_queued_packages, {
+      mod = self,
+      package_name = package_name,
+      resource_package = resource_package,
+      callback = callback,
+    })
+  end
+end
+
+function VMFMod:unload_package(package_name)
+  if vmf.check_wrong_argument_type(self, "unload_package", "package_name", package_name, "string") then
+    return
+  end
+
+  if not self:has_package_loaded(package_name) then
+    self:error("Package '%s' has not been loaded", package_name)
+    return
+  end
+
+  local resource_package = _loaded_packages[self][package_name]
+
+  resource_package:unload()
+  Mod.release_resource_package(resource_package)
+  _loaded_packages[self][package_name] = nil
+end
+
+function VMFMod:is_package_loading(package_name)
+  if vmf.check_wrong_argument_type(self, "is_package_loading", "package_name", package_name, "string") then
+    return
+  end
+
+  if _loading_package and _loading_package.mod == self and _loading_package.package_name == package_name then
+    return true
+  end
+
+  for _, queued_package in ipairs(_queued_packages) do
+    if queued_package.mod == self and queued_package.package_name == package_name then
+      return true
+    end
+  end
+
+  return false
+end
+
+function VMFMod:has_package_loaded(package_name)
+  if vmf.check_wrong_argument_type(self, "has_package_loaded", "package_name", package_name, "string") then
+    return
+  end
+
+  local loaded_packages = _loaded_packages[self]
+  return loaded_packages and loaded_packages[package_name] ~= nil
+end
+
+function vmf.update_package_manager()
+  local loading_package = _loading_package
+
+  if loading_package then
+    if loading_package.resource_package:has_loaded() then
+      _loaded_packages[loading_package.mod][loading_package.package_name] = loading_package.resource_package
+      _loading_package = nil
+
+      -- The callback has to be called last, so that any calls to `has_package_loaded` or `is_package_loading` return the correct value
+      loading_package.callback()
```

Comment:
Execution of any external mod functions should be wrapped in safe call in order to not crash VMF in case of error.
```suggestion
      vmf.safe_call_nr(loading_package.mod, {"'%s' package loaded callback", loading_package.package_name},
                        loading_package.callback, loading_package.package_name)
```
I also added passing package name to the callback function so the modder can identify for which package callback was loaded. Is this a good idea? What do you think? If not, just remove the last argument.